app/components/GMap.jsx

```diff
@@ -1,81 +1,120 @@
-        navigator.geolocation.getCurrentPosition( (position) => {
-          this.setState({
-            center: this.mapCenter(position.coords.latitude, position.coords.longitude)
-          });
-          this.moveMap("Got it!");
-        }, () => this.infoWindow.setContent("Couldn't find your location :("))
+      if (this.props.config && this.props.config.snapToUserLocation && navigator.geolocation) {
+        this.getUserLocation()
       } else {
         this.setState({
-          center: this.mapCenter(this.props.center.lat, this.props.center.lng)
+          center: this.mapCenter(this.props.config.initialCenter.lat, this.props.config.initialCenter.lng)
         })
       }
-      // create the map, marker and infoWindow after the component has
+      // create the map and markers after the component has
       // been rendered because we need to manipulate the DOM for Google =(
-      this.map = this.createMap(this.state.center);
-      this.marker = this.newMarker(this.state.center, this.map, this.props.markerImage);
-      this.infoWindow = this.newInfoWindow(this.map, this.marker, this.props.message);
-
-      // have to define google maps event listeners here too
-      // because we can't add listeners on the map until its created
-      google.maps.event.addListener(this.map, 'zoom_changed', () => this.handleZoomChange());
-      google.maps.event.addListener(this.marker, 'click', () => this.toggleInfoWindow())
+      this.map = this.createMap(this.props.config.initialCenter);
+      if (this.props.config && this.props.config.markers) {
+        this.markers = this.createMarkers(this.props.config.markers);
+        if (this.props.config.legend) {
+          this.createLegend(this.props.config.icons);
+        }
+      }
     }
   }
 
   // clean up event listeners when component unmounts
   componentDidUnMount() {
-    google.maps.event.clearListeners(map, 'zoom_changed')
+    google.maps.event.clearListeners(map, 'click')
+  }
+
+  createLegend(icons) {
+    let legend = this.refs.legend;
+    for (let key in icons) {
+      let type = icons[key];
+      let name = type.name;
+      let icon = type.image;
+      var div = document.createElement('div');
+      div.innerHTML = `<img src="${icon}"> ${name}`;
+      legend.appendChild(div);
+    }
+    this.map.controls[google.maps.ControlPosition.RIGHT_BOTTOM].push(legend);
   }
 
   createMap(center) {
     let mapOptions = {
-      zoom: this.state.zoom,
+      zoom: this.props.config.initialZoom,
       center: center,
-      mapTypeId: 'terrain'
     }
-    if (this.props.colors) {
-      mapOptions.styles = MapStyles(this.props.colors)
+    if (this.props.config && this.props.config.colors) {
+      mapOptions.styles = MapStyles(this.props.config.colors)
+      mapOptions.mapTypeId = 'terrain'
     }
     return new google.maps.Map(this.refs.mapCanvas, mapOptions)
   }
 
+  createMarkers(markers) {
+    markers.forEach( (marker) => {
+      if (this.props.config.icons) {
```

Comment:
That `var` was a scoping issue - because `let` and `const` are block-scoped, I couldn't access them outside the if/else. Your suggestion sidesteps the whole problem - that's a neat trick using `&&` when assigning `icon`!